### ISSUE TITLE
Unusual slow training speed

### ISSUE THREAD

**Issue by Fra-Tsuna** [prs-eth/Marigold#94]
Hi everyone,
first of all: really incredible work!

I have been trying to run the training from scratch. However, after setupping all the datasets in the required format and the files in the `config/` directory accordingly, the training speed appears unexpectedly slow. I'm talking about ~1 hour per iteration, and this is strange since we have comparable specs as the ones mentioned in the paper, and the claim is that it took 2.5 to completely train the model.

Do you have an idea about why it is taking so long? Also, could you please clarify what is the difference between epochs and iterations in your code?

Here more detailed information:

1. Script:
I'm launching the training with the following script:
`
python train.py --config config/train_marigold.yaml 
                --output_dir output/marigold 
                --base_data_dir $BASE_DATA_DIR 
                --base_ckpt_dir $BASE_CKPT_DIR 
                --add_datetime_prefix
`

2. Specs:
NVIDIA GeForce RTX 4090 with 24 GB of VRAM, 128 GB of RAM, CPU i9-13900KF

3. Logger output (up to 18 iter, but it is still going):
2024-07-01 15:51:12,900 - INFO -train.py - <module> >> device = cuda
 2024-07-01 15:51:12,905 - INFO -train.py - <module> >> Config saved to output/marigold/24_07_01-15_51_09-train_marigold/config.yaml 
 2024-07-01 15:51:13,093 - INFO -train.py - <module> >> Code snapshot saved to: output/marigold/24_07_01-15_51_09-train_marigold/code_snapshot.tar
 2024-07-01 15:51:13,093 - INFO -train.py - <module> >> Effective batch size: 32, accumulation steps: 16
 2024-07-01 15:51:13,099 - DEBUG -train.py - <module> >> Augmentation: 
 2024-07-01 15:51:13,331 - DEBUG -train.py - <module> >> Trainer: <class 'src.trainer.marigold_trainer.MarigoldTrainer'>
 2024-07-01 15:51:13,332 - INFO -marigold_trainer.py - _replace_unet_conv_in >> Unet conv_in layer is replaced
 2024-07-01 15:51:13,332 - INFO -marigold_trainer.py - _replace_unet_conv_in >> Unet config is updated
 2024-07-01 15:51:14,026 - INFO -marigold_trainer.py - train >> Start training
 2024-07-01 15:51:14,519 - DEBUG -marigold_trainer.py - train >> epoch: 1
 2024-07-01 16:03:52,834 - INFO -marigold_trainer.py - _get_next_seed >> Global seed sequence is generated, length=480000
 2024-07-01 16:57:35,587 - INFO -marigold_trainer.py - train >> iter     1 (epoch  1): loss=0.84598
 2024-07-01 17:58:02,353 - INFO -marigold_trainer.py - train >> iter     2 (epoch  1): loss=0.66081
 2024-07-01 18:51:36,170 - INFO -marigold_trainer.py - train >> iter     3 (epoch  1): loss=0.73767
 2024-07-01 19:52:43,342 - INFO -marigold_trainer.py - train >> iter     4 (epoch  1): loss=0.77796
 2024-07-01 20:43:07,365 - INFO -marigold_trainer.py - train >> iter     5 (epoch  1): loss=0.73497
 2024-07-01 21:39:41,658 - INFO -marigold_trainer.py - train >> iter     6 (epoch  1): loss=0.74770
 2024-07-01 22:34:36,260 - INFO -marigold_trainer.py - train >> iter     7 (epoch  1): loss=0.66351
 2024-07-01 23:31:44,398 - INFO -marigold_trainer.py - train >> iter     8 (epoch  1): loss=0.66380
 2024-07-02 00:28:58,972 - INFO -marigold_trainer.py - train >> iter     9 (epoch  1): loss=0.74787
 2024-07-02 01:27:05,695 - INFO -marigold_trainer.py - train >> iter    10 (epoch  1): loss=0.75394
 2024-07-02 02:22:03,929 - INFO -marigold_trainer.py - train >> iter    11 (epoch  1): loss=0.68349
 2024-07-02 03:17:13,801 - INFO -marigold_trainer.py - train >> iter    12 (epoch  1): loss=0.88774
 2024-07-02 04:08:37,226 - INFO -marigold_trainer.py - train >> iter    13 (epoch  1): loss=0.70726
 2024-07-02 05:04:02,808 - INFO -marigold_trainer.py - train >> iter    14 (epoch  1): loss=0.67777
 2024-07-02 06:04:16,033 - INFO -marigold_trainer.py - train >> iter    15 (epoch  1): loss=0.67077
 2024-07-02 07:00:04,422 - INFO -marigold_trainer.py - train >> iter    16 (epoch  1): loss=0.66342
 2024-07-02 07:58:48,317 - INFO -marigold_trainer.py - train >> iter    17 (epoch  1): loss=0.55943
 2024-07-02 08:53:59,042 - INFO -marigold_trainer.py - train >> iter    18 (epoch  1): loss=0.72592

4. Data format:
All training, testing and validation datasets are in the `data/` directory in the following structure:
`data/name_dataset/dataset.tar` 
according to the configuration files, and they are all on SSD.

 If you need more details I will provide them to you. Thanks in advance for your help.

**Comment by markkua** [comment]
Hi, I would check if the it is using GPU or CPU (by checking `device` variable [here](https://github.com/prs-eth/Marigold/blob/f74115261b67b59fb536994d0413f64d69af65c5/train.py#L195) and monitoring GPU utilization). 

To eliminate the possibility of slow data loading, you can also run a unit test on the dataloader to check its speed.

**Comment by markkua** [comment]
How is the GPU utilization? There seems to be something blocking the computation even though the model is loaded into the VRAM.
I would also check the package versions and try using more stable versions (e.g. using diffusers==0.25.0)

Regarding epochs vs. iteration:
- 1 iteration: optimizer.step() is called once. In the default setting, it corresponds to x16 accumulation steps.
- 1 epoch: dataset is passed through the model once (I don't use this concept in the code, so simply set it to a large enough value)

**Comment by water221** [comment]
> Hi everyone, first of all: really incredible work!
> 
> I have been trying to run the training from scratch. However, after setupping all the datasets in the required format and the files in the `config/` directory accordingly, the training speed appears unexpectedly slow. I'm talking about ~1 hour per iteration, and this is strange since we have comparable specs as the ones mentioned in the paper, and the claim is that it took 2.5 to completely train the model.
> 
> Do you have an idea about why it is taking so long? Also, could you please clarify what is the difference between epochs and iterations in your code?
> 
> Here more detailed information:
> 
> 1. Script:
>    I'm launching the training with the following script:
>    `python train.py --config config/train_marigold.yaml  --output_dir output/marigold  --base_data_dir $BASE_DATA_DIR  --base_ckpt_dir $BASE_CKPT_DIR  --add_datetime_prefix`
> 2. Specs:
>    NVIDIA GeForce RTX 4090 with 24 GB of VRAM, 128 GB of RAM, CPU i9-13900KF
> 3. Logger output (up to 18 iter, but it is still going):
>    2024-07-01 15:51:12,900 - INFO -train.py -  >> device = cuda
>    2024-07-01 15:51:12,905 - INFO -train.py -  >> Config saved to output/marigold/24_07_01-15_51_09-train_marigold/config.yaml
>    2024-07-01 15:51:13,093 - INFO -train.py -  >> Code snapshot saved to: output/marigold/24_07_01-15_51_09-train_marigold/code_snapshot.tar
>    2024-07-01 15:51:13,093 - INFO -train.py -  >> Effective batch size: 32, accumulation steps: 16
>    2024-07-01 15:51:13,099 - DEBUG -train.py -  >> Augmentation:
>    2024-07-01 15:51:13,331 - DEBUG -train.py -  >> Trainer: <class 'src.trainer.marigold_trainer.MarigoldTrainer'>
>    2024-07-01 15:51:13,332 - INFO -marigold_trainer.py - _replace_unet_conv_in >> Unet conv_in layer is replaced
>    2024-07-01 15:51:13,332 - INFO -marigold_trainer.py - _replace_unet_conv_in >> Unet config is updated
>    2024-07-01 15:51:14,026 - INFO -marigold_trainer.py - train >> Start training
>    2024-07-01 15:51:14,519 - DEBUG -marigold_trainer.py - train >> epoch: 1
>    2024-07-01 16:03:52,834 - INFO -marigold_trainer.py - _get_next_seed >> Global seed sequence is generated, length=480000
>    2024-07-01 16:57:35,587 - INFO -marigold_trainer.py - train >> iter     1 (epoch  1): loss=0.84598
>    2024-07-01 17:58:02,353 - INFO -marigold_trainer.py - train >> iter     2 (epoch  1): loss=0.66081
>    2024-07-01 18:51:36,170 - INFO -marigold_trainer.py - train >> iter     3 (epoch  1): loss=0.73767
>    2024-07-01 19:52:43,342 - INFO -marigold_trainer.py - train >> iter     4 (epoch  1): loss=0.77796
>    2024-07-01 20:43:07,365 - INFO -marigold_trainer.py - train >> iter     5 (epoch  1): loss=0.73497
>    2024-07-01 21:39:41,658 - INFO -marigold_trainer.py - train >> iter     6 (epoch  1): loss=0.74770
>    2024-07-01 22:34:36,260 - INFO -marigold_trainer.py - train >> iter     7 (epoch  1): loss=0.66351
>    2024-07-01 23:31:44,398 - INFO -marigold_trainer.py - train >> iter     8 (epoch  1): loss=0.66380
>    2024-07-02 00:28:58,972 - INFO -marigold_trainer.py - train >> iter     9 (epoch  1): loss=0.74787
>    2024-07-02 01:27:05,695 - INFO -marigold_trainer.py - train >> iter    10 (epoch  1): loss=0.75394
>    2024-07-02 02:22:03,929 - INFO -marigold_trainer.py - train >> iter    11 (epoch  1): loss=0.68349
>    2024-07-02 03:17:13,801 - INFO -marigold_trainer.py - train >> iter    12 (epoch  1): loss=0.88774
>    2024-07-02 04:08:37,226 - INFO -marigold_trainer.py - train >> iter    13 (epoch  1): loss=0.70726
>    2024-07-02 05:04:02,808 - INFO -marigold_trainer.py - train >> iter    14 (epoch  1): loss=0.67777
>    2024-07-02 06:04:16,033 - INFO -marigold_trainer.py - train >> iter    15 (epoch  1): loss=0.67077
>    2024-07-02 07:00:04,422 - INFO -marigold_trainer.py - train >> iter    16 (epoch  1): loss=0.66342
>    2024-07-02 07:58:48,317 - INFO -marigold_trainer.py - train >> iter    17 (epoch  1): loss=0.55943
>    2024-07-02 08:53:59,042 - INFO -marigold_trainer.py - train >> iter    18 (epoch  1): loss=0.72592
> 4. Data format:
>    All training, testing and validation datasets are in the `data/` directory in the following structure:
>    `data/name_dataset/dataset.tar`
>    according to the configuration files, and they are all on SSD.
> 
> If you need more details I will provide them to you. Thanks in advance for your help.

Excuse me, can you explain whether your ```BASE_CKPT_DIR``` is set to the directory of ```marigold-v1-0``` or the directory of ```768-v-ema.ckpt```. I am currently reproducing this paper, but after I started using the command of python train.py--config/train_marigold.yaml --no_wandb, the log only contained the content in the picture, and then there was no other action or output
![image](https://github.com/prs-eth/Marigold/assets/52553731/da50543e-654d-4f1f-8776-64dd09564d5f)

**Comment by Fra-Tsuna** [comment]
We identified the problem: the DataLoader's extraction of files from a tar archive. This method introduced excessive overhead, causing the model to load into the GPU but not utilize it properly (which apparently is a known bug/behaviour when there is too much overhead). By unzipping the folders and loading the images directly, we achieved the correct training speed. Thank you for your assistance.

@water221 the `BASE_CKPT_DIR` variable contains the path to the (local) official `stable-diffusion-2` repository cloned from hugging-face. I will close this Issue now, if you have more trouble I suggest you to open another Issue.

**Comment by water221** [comment]
Thanks for your answer, I will try it

**Comment by YacineDeghaies** [comment]
> Thanks for your answer, I will try it

under /checkpoint
do this: git clone https://huggingface.co/stabilityai/stable-diffusion-2

**Comment by water221** [comment]
> > Thanks for your answer, I will try it
> 
> under /checkpoint do this: git clone https://huggingface.co/stabilityai/stable-diffusion-2

Thanks for the tip

**Comment by water221** [comment]
> We identified the problem: the DataLoader's extraction of files from a tar archive. This method introduced excessive overhead, causing the model to load into the GPU but not utilize it properly (which apparently is a known bug/behaviour when there is too much overhead). By unzipping the folders and loading the images directly, we achieved the correct training speed. Thank you for your assistance.
> 
@Fra-Tsuna Hello, I am a beginner in this field. Recently, I encountered the same problem as you during training. After seeing that you have directly loaded images by unzipping the folder to achieve normal training speed, I wonder if you can explain how to make changes in the code to realize direct loading of images. Do you want to make changes in the config file, or make corresponding changes in the train.py file. I would be grateful if you could show how you made changes
